### ISSUE TITLE
Git compatibility for exp names

### ISSUE THREAD

**Issue by dmpetrov** [iterative/dvc#5737]
```bash
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created  ┃ step ┃   loss ┃ accuracy ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -        │    9 │ 1.8085 │  0.52762 │ … │ 15     │ … │
│ master                │ 09:27 AM │    - │     15 │  2.1e-07 │ … │ 0.124  │ … │
│ │ ╓ exp-dba5c         │ 10:42 AM │    9 │ 1.8085 │  0.52762 │ … │ 15     │ … │
│ │ ╟ 00e68eb           │ 10:42 AM │    8 │  1.857 │  0.48758 │ … │ 15     │ … │
│ │ ╟ cda7883 (3263dc2) │ 10:42 AM │    7 │ 1.9092 │  0.44263 │ … │ 15     │ … │
│ │ ╓ exp-59c7a         │ 10:41 AM │    6 │ 1.9522 │  0.42667 │ … │ 15     │ … │
│ │ ╟ d721197           │ 10:41 AM │    5 │ 2.0103 │  0.36575 │ … │ 15     │ … │
│ │ ╟ 0d4e1a7           │ 10:40 AM │    4 │ 2.0717 │  0.30868 │ … │ 15     │ … │
│ │ ╟ 312a0e5           │ 10:40 AM │    3 │ 2.1351 │   0.2546 │ … │ 15     │ … │
│ │ ╟ a00b24c           │ 10:40 AM │    2 │  2.202 │  0.20255 │ … │ 15     │ … │
│ │ ╟ a371c0e           │ 10:39 AM │    1 │ 2.2746 │  0.15662 │ … │ 15     │ … │
│ ├─╨ 3da11aa           │ 10:39 AM │    0 │ 2.3516 │  0.12398 │ … │ 15     │ … │
└───────────────────────┴──────────┴──────┴────────┴──────────┴───┴────────┴───┘
```

```bash
# works just great!
$ git diff 00e68eb 0d4e1a7 -- params.yaml
diff --git a/params.yaml b/params.yaml
index 8f6f5fb..fedaf92 100644
...
...

# These two do not work :( 
$ git diff exp-dba5c 0d4e1a7 -- params.yaml 
fatal: bad revision 'exp-dba5c'
$ git diff dba5c 0d4e1a7 -- params.yaml
fatal: bad revision 'dba5c'
```

I'd love to have Git compatibility for exp names like `exp-dba5c` or at least the suffixes `dba5c`.

EDIT:
The same should ideally work for `dvc metrics` & `dvc params`.

**Comment by dberenbaum** [comment]
@pmrowla can correct me if I'm wrong, but I think this is kind of the point of the custom references. They are just like Git branches, except that Git mostly ignores them. Which Git commands should recognize these refs? I think it's obvious we don't want them to show up in some Git commands, like `git branch`, and I'm not sure if it's worthwhile to define which Git commands should support them (not to mention I'm not sure how feasible this is to implement).

> The same should ideally work for `dvc metrics` & `dvc params`.

These should work and do in the basic tests I tried. Follow up if you find they don't work.

**Comment by pmrowla** [comment]
DVC commands already work with exp names, but git commands will not, and there's not really anything we can do there. For the high level porcelain commands, the git CLI is hard coded to only resolve short ref names for tags (`refs/tags`) and branches (`refs/heads`).

To use commands like `git diff` with experiments, it will only work if you pass the commit SHA (or the full path to the ref name).
You can get the commit SHA instead of experiment name using `dvc exp show --sha`

(You can also get commit both commit SHA and the full path to a ref using the lower level plumbing command `git show-ref exp-dba5c`)

---

For `diff` specifically, one potential workaround would be for us to add a flag to include the git (code) diff output in `dvc exp diff`.

**Comment by shcheklein** [comment]
A suggestion from the top of my head - how about creating a wrapper (proxy) - `dvc git ...` that will do the translation and pass the right references down to the Git backend? It's not as convenient but seems a minor thing to remember when you deal with it everyday.

**Comment by pmrowla** [comment]
That seems like a case where we would need to alias every git porcelain command though? Some git commands take refnames as `--option=...` fields, some take them as regular (`cmd <arg1> <arg2> ...`) arguments, and some commands want explicit names (and not resolved commit SHAs), so our wrapper would have to know the semantics for each of these commands.

**Comment by dmpetrov** [comment]
Oh, right. It won't be easy to make this `exp-X` actionable. `dvc exp show --sha` solves the issue and seems like a better workaround than a git wrapper:
```bash
$ dvc exp show --no-pager --sha
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃   loss ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -            │   12 │ 1.6681 │ … │ 15     │ … │
│ 10dd615               │ Mar 30, 2021 │    - │     15 │ … │ 0.124  │ … │
│ │ ╓ d1db676           │ Mar 30, 2021 │   12 │ 1.6681 │ … │ 15     │ … │
│ │ ╟ 9a6aa59           │ Mar 30, 2021 │   11 │ 1.7136 │ … │ 15     │ … │
│ │ ╟ 32c62a0 (99a6731) │ Mar 30, 2021 │   10 │ 1.7597 │ … │ 15     │ … │
│ │ ╓ 99a6731           │ Mar 30, 2021 │    9 │ 1.8085 │ … │ 15     │ … │
│ │ ╟ 00e68eb           │ Mar 30, 2021 │    8 │  1.857 │ … │ 15     │ … │
│ │ ╟ cda7883 (3263dc2) │ Mar 30, 2021 │    7 │ 1.9092 │ … │ 15     │ … │
│ │ ╓ 3263dc2           │ Mar 30, 2021 │    6 │ 1.9522 │ … │ 15     │ … │
│ │ ╟ d721197           │ Mar 30, 2021 │    5 │ 2.0103 │ … │ 15     │ … │
│ │ ╟ 0d4e1a7           │ Mar 30, 2021 │    4 │ 2.0717 │ … │ 15     │ … │
│ │ ╟ 312a0e5           │ Mar 30, 2021 │    3 │ 2.1351 │ … │ 15     │ … │
│ │ ╟ a00b24c           │ Mar 30, 2021 │    2 │  2.202 │ … │ 15     │ … │
│ │ ╟ a371c0e           │ Mar 30, 2021 │    1 │ 2.2746 │ … │ 15     │ … │
│ ├─╨ 3da11aa           │ Mar 30, 2021 │    0 │ 2.3516 │ … │ 15     │ … │
└───────────────────────┴──────────────┴──────┴────────┴───┴────────┴───┘
```

It would be great to see the sha-s in the default table. Is it possible to include sha in the table?

I see two options:
1. add sha next to experiment:
```bash
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃   loss ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -            │   12 │ 1.6681 │ … │ 15     │ … │
│ master                │ Mar 30, 2021 │    - │     15 │ … │ 0.124  │ … │
│ │ ╓ d1db676(exp-5d1dc)│ Mar 30, 2021 │   12 │ 1.6681 │ … │ 15     │ … │
│ │ ╟ 9a6aa59           │ Mar 30, 2021 │   11 │ 1.7136 │ … │ 15     │ … │
│ │ ╟ 32c62a0 (99a6731) │ Mar 30, 2021 │   10 │ 1.7597 │ … │ 15     │ … │
│ │ ╓ 99a6731(exp-dba5c)│ Mar 30, 2021 │    9 │ 1.8085 │ … │ 15     │ … │
│ │ ╟ 00e68eb           │ Mar 30, 2021 │    8 │  1.857 │ … │ 15     │ … │
│ │ ╟ cda7883 (3263dc2) │ Mar 30, 2021 │    7 │ 1.9092 │ … │ 15     │ … │
│ │ ╓ 3263dc2(exp-59c7a)│ Mar 30, 2021 │    6 │ 1.9522 │ … │ 15     │ … │
│ │ ╟ d721197           │ Mar 30, 2021 │    5 │ 2.0103 │ … │ 15     │ … │
│ │ ╟ 0d4e1a7           │ Mar 30, 2021 │    4 │ 2.0717 │ … │ 15     │ … │
│ │ ╟ 312a0e5           │ Mar 30, 2021 │    3 │ 2.1351 │ … │ 15     │ … │
│ │ ╟ a00b24c           │ Mar 30, 2021 │    2 │  2.202 │ … │ 15     │ … │
│ │ ╟ a371c0e           │ Mar 30, 2021 │    1 │ 2.2746 │ … │ 15     │ … │
│ ├─╨ 3da11aa           │ Mar 30, 2021 │    0 │ 2.3516 │ … │ 15     │ … │
└───────────────────────┴──────────────┴──────┴────────┴───┴────────┴───┘
```

2. Use sha in the table but add a separate line per exp :
```bash
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃   loss ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -            │   12 │ 1.6681 │ … │ 15     │ … │
│ master                │ Mar 30, 2021 │    - │     15 │ … │ 0.124  │ … │
│ │ ╓ exp-5d1dc         │              │      │        │ … │        │ … │
│ │ ╟ d1db676           │ Mar 30, 2021 │   12 │ 1.6681 │ … │ 15     │ … │
│ │ ╟ 9a6aa59           │ Mar 30, 2021 │   11 │ 1.7136 │ … │ 15     │ … │
│ │ ╟ 32c62a0 (99a6731) │ Mar 30, 2021 │   10 │ 1.7597 │ … │ 15     │ … │
│ │ ╓ exp-dba5c         │              │      │        │ … │        │ … │
│ │ ╟ 99a6731           │ Mar 30, 2021 │    9 │ 1.8085 │ … │ 15     │ … │
│ │ ╟ 00e68eb           │ Mar 30, 2021 │    8 │  1.857 │ … │ 15     │ … │
│ │ ╟ cda7883 (3263dc2) │ Mar 30, 2021 │    7 │ 1.9092 │ … │ 15     │ … │
│ │ ╓ exp-59c7a         │              │      │        │ … │        │ … │
│ │ ╟ 3263dc2           │ Mar 30, 2021 │    6 │ 1.9522 │ … │ 15     │ … │
│ │ ╟ d721197           │ Mar 30, 2021 │    5 │ 2.0103 │ … │ 15     │ … │
│ │ ╟ 0d4e1a7           │ Mar 30, 2021 │    4 │ 2.0717 │ … │ 15     │ … │
│ │ ╟ 312a0e5           │ Mar 30, 2021 │    3 │ 2.1351 │ … │ 15     │ … │
│ │ ╟ a00b24c           │ Mar 30, 2021 │    2 │  2.202 │ … │ 15     │ … │
│ │ ╟ a371c0e           │ Mar 30, 2021 │    1 │ 2.2746 │ … │ 15     │ … │
│ ├─╨ 3da11aa           │ Mar 30, 2021 │    0 │ 2.3516 │ … │ 15     │ … │
└───────────────────────┴──────────────┴──────┴────────┴───┴────────┴───┘
```

However, the 2nd approach is not the best solution for regular exps - we will need to duplicate every single line:
```bash
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━┓
┃ Experiment    ┃ Created      ┃ avg_prec ┃ roc_auc ┃ prepare.split ┃ … ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━┩
│ workspace     │ -            │   0.5795 │ 0.95942 │ 0.2           │ … │
│ master        │ Mar 01, 2021 │  0.60405 │  0.9608 │ 0.2           │ … │
│ ├── exp-d4629 │ Mar 30, 2021 │   0.5795 │ 0.95942 │ 0.2           │ … │
│ ├── exp-e37cb │ Mar 30, 2021 │  0.60462 │  0.9596 │ 0.2           │ … │
│ ├── exp-f6369 │ Mar 30, 2021 │   0.6044 │ 0.95967 │ 0.2           │ … │
│ ├── exp-44136 │ Mar 30, 2021 │  0.60405 │  0.9608 │ 0.2           │ … │
│ └── exp-0e1ef │ Mar 30, 2021 │  0.60405 │  0.9608 │ 0.2           │ … │
└───────────────┴──────────────┴──────────┴─────────┴───────────────┴───┘
```

It seems like adding sha to the table is a good solution. WDYH?

**Comment by pmrowla** [comment]
Including the SHA's in the table by default is simple enough, we just need to decide how we want to display them. The one thing to keep in mind is that we already use parentheses to denote parents for branched/modified checkpoint runs like in
```
│ │ ╟ 32c62a0 (99a6731)
```

**Comment by dberenbaum** [comment]
If dvc shows sha by default for experiments without checkpoints, how do I see the experiment names instead?

IMHO the current solution of the `--sha` flag is pretty functional. I'd vote to leave as is (and generally I'd like to start being more protective about experiments UI requests so @pmrowla can start on other features 😄 ).

**Comment by dmpetrov** [comment]
>> is to show only sha-s in regular experiments (not checkpoints).

> If dvc shows sha by default for experiments without checkpoints, how do I see the experiment names instead?

It was only about regular experiments where a name might be required only when a user specifies it (not a usual case).


> This option ^^ looks the best so far. Is there anything I'm missing here?

Actually, it is not 😄  Because it will lead to duplication for the regular experiments.


> IMHO the current solution of the `--sha` flag is pretty functional.

Yes, but only if we make this a default behavior. Otherwise, we are pushing users to remember 2 concepts and an additional option. It will create overhead in slides (I already have this) and docs. If we show sha-s all the time, we still need to show the names when they might be needed:
1. regular experiment - when user changed the name.
2. checkpoints

Showing sha in `d1db676(exp-5d1dc)` format might be a better choice because we should avoid unnecessary duplication for the regular exp table (1 sha for each experiment).

A separate question. Prefix `exp-` might be not the best one since we already abusing the **experiment** term and creating dumplication in links refs/**exps**/10/dd6159b1946c31fb4c496415287f7cad5fcacc/**exp**-59c7a. Does it make sense to change `exp-` prefixt to something like `n-` (name, and shorter)?

**Comment by dberenbaum** [comment]
> Showing sha in `d1db676(exp-5d1dc)` format might be a better choice because we should avoid unnecessary duplication for the regular exp table (1 sha for each experiment).

```
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃   loss ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -            │   12 │ 1.6681 │ … │ 15     │ … │
│ master                │ Mar 30, 2021 │    - │     15 │ … │ 0.124  │ … │
│ │ ╓ d1db676(exp-5d1dc)│ Mar 30, 2021 │   12 │ 1.6681 │ … │ 15     │ … │
│ │ ╟ 9a6aa59           │ Mar 30, 2021 │   11 │ 1.7136 │ … │ 15     │ … │
│ │ ╟ 32c62a0 (99a6731) │ Mar 30, 2021 │   10 │ 1.7597 │ … │ 15     │ … │
│ │ ╓ 99a6731(exp-dba5c)│ Mar 30, 2021 │    9 │ 1.8085 │ … │ 15     │ … │
│ │ ╟ 00e68eb           │ Mar 30, 2021 │    8 │  1.857 │ … │ 15     │ … │
│ │ ╟ cda7883 (3263dc2) │ Mar 30, 2021 │    7 │ 1.9092 │ … │ 15     │ … │
│ │ ╓ 3263dc2(exp-59c7a)│ Mar 30, 2021 │    6 │ 1.9522 │ … │ 15     │ … │
│ │ ╟ d721197           │ Mar 30, 2021 │    5 │ 2.0103 │ … │ 15     │ … │
│ │ ╟ 0d4e1a7           │ Mar 30, 2021 │    4 │ 2.0717 │ … │ 15     │ … │
│ │ ╟ 312a0e5           │ Mar 30, 2021 │    3 │ 2.1351 │ … │ 15     │ … │
│ │ ╟ a00b24c           │ Mar 30, 2021 │    2 │  2.202 │ … │ 15     │ … │
│ │ ╟ a371c0e           │ Mar 30, 2021 │    1 │ 2.2746 │ … │ 15     │ … │
│ ├─╨ 3da11aa           │ Mar 30, 2021 │    0 │ 2.3516 │ … │ 15     │ … │
└───────────────────────┴──────────────┴──────┴────────┴───┴────────┴───┘
```

You mean this table format, right?

One concern is the amount of info stored in the first column. In this example, it contains info about branches/tree structure, relationships/ordering between checkpoints, experiment names, commit shas, and parent commits for modified checkpoint runs. Having a concise UI is great, but I wonder if it makes more sense at some point to split into multiple columns (for example, commit sha and experiment name columns). Personally, I don't think I would understand what's in the parentheses if I wasn't working on these features.

> Does it make sense to change `exp-` prefixt to something like `n-` (name, and shorter)?

Hm, I think `exp-` is clear because it is used a lot elsewhere, and `n-` might be less obvious. I also think the length of the random part of the string matters a lot more than `exp-` vs `n-` since the random part is harder to remember.

**Comment by dmpetrov** [comment]
> You mean this table format, right?

Right.

> Personally, I don't think I would understand what's in the parentheses if I wasn't working on these features.

Yes. The confusion comes from experiments `(exp-5d1dc)` and parents `(99a6731)`. If one of these can be extracted in an (optional?) column while we can keep another one. Making two columns might be too much.

So far a proper solution might looked like:
1. Show sha-s `9a6aa59 ` for all the items in the table.
2. Show name in parentheses `99a6731(exp-dba5c)` for checkpoints head. For regular experiments - only if exp name is not default.
3. Do not show parents. Exptrtact them in a separate optional column. Use `--parent` to enable the column.

**Comment by dberenbaum** [comment]
Regardless of the exact formatting, can we first decide if sha is an obviously better default than exp name? For Git commands, sha is preferable, but many of the dvc exp commands (apply, branch, list, push, pull, remove) require exp name.

**Comment by dmpetrov** [comment]
> can we first decide if sha is an obviously better default than exp name?  

Right. It seems like she might be a better default since it is actionable for both Git and DVC.

> ... many of the dvc exp commands (apply, branch, list, push, pull, remove) require exp name.

I guess, it can be supported in DVC if its not implemented yet. @pmrowla are there any cases when sha cannot be used, only exp name?

What I found so far:
```bash
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃     loss ┃ accuracy ┃ val_loss ┃ val_accuracy ┃ epochs ┃ lr      ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━┩
│ workspace             │ -            │    7 │ 0.054196 │  0.98373 │ 0.067046 │       0.9784 │ 15     │ 0.0005  │ … │
│ master                │ Mar 30, 2021 │    - │        - │        - │        - │            - │ 15     │ 2.1e-07 │ … │
│ │ ╓ exp-70928         │ 07:47 PM     │    7 │ 0.054196 │  0.98373 │ 0.067046 │       0.9784 │ 15     │ 0.0005  │ … │
│ │ ╟ 47b34d6           │ 07:47 PM     │    6 │ 0.074431 │  0.97763 │ 0.076507 │        0.975 │ 15     │ 0.0005  │ … │
│ │ ╟ 5471a57           │ 07:47 PM     │    5 │  0.11097 │  0.96748 │ 0.088773 │       0.9729 │ 15     │ 0.0005  │ … │
│ │ ╟ e9ebc70 (7af692d) │ 07:47 PM     │    4 │  0.25394 │  0.92745 │  0.12504 │       0.9632 │ 15     │ 0.0005  │ … │
│ │ ╓ exp-8b809         │ 07:47 PM     │    3 │   2.1693 │   0.2806 │   2.1272 │         0.35 │ 15     │ 2.1e-07 │ … │
│ │ ╟ c9cd463           │ 07:47 PM     │    2 │   2.2433 │  0.21087 │   2.2005 │       0.2653 │ 15     │ 2.1e-07 │ … │
│ │ ╟ a7a30b4           │ 07:47 PM     │    1 │   2.3233 │  0.15603 │   2.2794 │       0.1925 │ 15     │ 2.1e-07 │ … │
│ ├─╨ a1fddb3           │ 07:47 PM     │    0 │   2.4123 │   0.1174 │   2.3657 │       0.1307 │ 15     │ 2.1e-07 │ … │
└───────────────────────┴──────────────┴──────┴──────────┴──────────┴──────────┴──────────────┴────────┴─────────┴───┘
```

```bash
$ dvc exp apply c9cd463
ERROR: Ambiguous commit 'c9cd463' belongs to multiple experiment branches.
```

While the other sha works just fine:
```bash
$ dvc exp apply 5471a57
Changes for experiment '5471a57' have been applied to your current workspace.
```

What is the difference between `5471a57` and `c9cd463`?

**Comment by pmrowla** [comment]
Experiment names are required for commands where git branch names would also be required - `git push/pull`, `git branch`, `git show-ref/for-each-ref` (for listing refs) all require ref names and not SHAs, so `dvc exp push/pull`, `exp branch`, `exp list` also require names and not SHAs.

Everything else should allow either SHAs or names.

> ```
> $ dvc exp apply c9cd463
> ERROR: Ambiguous commit 'c9cd463' belongs to multiple experiment branches.
> ```

This looks like a bug for `exp apply` and should go into a separate issue.

The command is failing an internal check where we try to determine what experiment Git commit `c9cd463` belongs to. In this example, the specified commit belongs to the shared git history of both `exp-8b809` and `exp-70928` (since the second exp branches off of the first one). Commits belonging to more than one experiment is an error condition for certain DVC functionality where we can only do an operation on a single experiment (and as a result DVC cannot guess which experiment the user meant unless the user explicitly uses the experiment name), but for `apply` in particular it should probably not error out here.

The second `apply` passes because commit `5471a57` only belongs to the second experiment's history (since it came after the branch point). In this case DVC is able to determine that the user meant "a commit inside `exp-70928`.

**Comment by dmpetrov** [comment]
@pmrowla thank you for the clarification!

> so `dvc exp push/pull`, `exp branch`, `exp list` also require names and not SHAs.

Is there a way to translate sha to exp? If not - it might be better to keep all exp names by default (like `d1db676(exp-5d1dc)`) even for not-checkpoint exps.

> However, what we should maybe be doing here is assuming that the user intends to apply the intermediate commit and then start a new branch from that point? (this discussion should probably be moved into a new issue)

`exp apply` & file modify & `exp run` - works just fine for a sha within a later experiment (when there is no ambiguous-commit issue). However, `exp branch sha` creates branch on the "head" of experiment. Is it expected? Does it make sense to change this?

Is there a way to avoid ambiguous-commit? Even if a commit is presented in several experiments, DVC can take any of these. Can't it?

**Comment by pmrowla** [comment]
> Is there a way to translate sha to exp? If not - it might be better to keep all exp names by default (like `d1db676(exp-5d1dc)`) even for not-checkpoint exps.

> However, `exp branch sha` creates branch on the "head" of experiment. Is it expected? Does it make sense to change this?

These 2 things are related, we can translate a sha to exp (as long as that sha only belongs to a single experiment), which is what happens in `exp branch sha`. Since the `exp branch` command really needs a name and not any arbitrary commit, we assume that what the user meant is the experiment containing sha, and then use that entire experiment (all the way up to the "head" of the exp) as the branch.

> exp apply & file modify & exp run - works just fine for a sha within a later experiment (when there is no ambiguous-commit issue)

> Is there a way to avoid ambiguous-commit? Even if a commit is presented in several experiments, DVC can take any of these. Can't it?

For `exp apply` it should probably accept any commit, but we have to determine the behavior for `exp run` in that case.

Normally, with `exp apply; exp run` for a sha that belongs to a single experiment (no ambiguous-commit issue), we resume that experiment from the "head". We cannot do this for the ambiguous commit issue, since there is more than one "head" for us to choose from.

But what we could do is create new experiment branch entirely, starting from that (ambiguous) commit.

**Comment by dmpetrov** [comment]
> Normally, with `exp apply; exp run` for a sha that belongs to a single experiment (no ambiguous-commit issue), we resume that experiment from the "head".

Could you please double-check this? To me, it seems to work from sha, not head. It looks like correct behavior.
```bash
## Apply intermediate checkpoint from the head of exp:
$ dvc exp apply a882855
Changes for experiment 'a882855' have been applied to your current workspace.
$ vi params.yaml    # MODIFY PARAMS
$ dvc exp run # AND CANCEL

## RESULT: training did not start from exp head (see accuracy metrics):
$ dvc exp show --no-pager
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━┳━━━━━━━━┳━━━┓
┃ Experiment            ┃ Created      ┃ step ┃   loss ┃ accuracy ┃ val_loss ┃ … ┃ epochs ┃ … ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━╇━━━━━━━━╇━━━┩
│ workspace             │ -            │   10 │ 1.7123 │   0.6156 │   1.6573 │ … │ 15     │ … │
│ master                │ Mar 30, 2021 │    - │      - │        - │        - │ - │ 15     │ … │
│ │ ╓ exp-57b0b         │ 08:54 PM     │   10 │ 1.7123 │   0.6156 │   1.6573 │ … │ 15     │ … │
│ │ ╟ 43c45a3           │ 08:54 PM     │    9 │ 1.7758 │   0.5747 │   1.7215 │ … │ 15     │ … │
│ │ ╟ 9166ef6           │ 08:54 PM     │    8 │ 1.8424 │  0.52585 │    1.788 │ … │ 15     │ … │
│ │ ╟ 8f5e0ee (a882855) │ 08:54 PM     │    7 │  1.913 │  0.46385 │   1.8571 │ … │ 15     │ … │
│ │ ╓ b-acc50           │ 08:53 PM     │   15 │ 1.5277 │   0.6973 │   1.4786 │ … │ 15     │ … │
│ │ ╟ 8ec1980           │ 08:53 PM     │   14 │ 1.5712 │  0.68122 │   1.5228 │ … │ 15     │ … │
│ │ ╟ add6b75           │ 08:52 PM     │   13 │ 1.6168 │  0.66382 │   1.5685 │ … │ 15     │ … │
│ │ ╟ bb66d7a           │ 08:52 PM     │   12 │ 1.6631 │  0.64097 │   1.6155 │ … │ 15     │ … │
│ │ ╟ c8d566c           │ 08:52 PM     │   11 │ 1.7112 │  0.61632 │   1.6639 │ … │ 15     │ … │
│ │ ╟ 63b368a           │ 08:52 PM     │   10 │  1.761 │  0.58572 │   1.7139 │ … │ 15     │ … │
│ │ ╟ fb154f7           │ 08:52 PM     │    9 │ 1.8125 │  0.54942 │   1.7653 │ … │ 15     │ … │
│ │ ╟ 1fec0b0           │ 08:52 PM     │    8 │ 1.8651 │  0.50863 │   1.8182 │ … │ 15     │ … │
│ │ ╟ 9c907ab           │ 08:52 PM     │    7 │ 1.9199 │  0.45958 │   1.8728 │ … │ 15     │ … │
│ │ ╟ a882855           │ 08:52 PM     │    6 │ 1.9755 │  0.40935 │   1.9292 │ … │ 15     │ … │
│ │ ╟ 3a4de2a           │ 08:52 PM     │    5 │ 2.0344 │  0.35383 │   1.9874 │ … │ 15     │ … │
│ │ ╟ 19f84e3           │ 08:52 PM     │    4 │ 2.0948 │   0.2938 │   2.0475 │ … │ 15     │ … │
│ │ ╟ bdde1fc (ab110cc) │ 08:52 PM     │    3 │ 2.1564 │  0.23527 │     2.11 │ … │ 15     │ … │
│ │ ╓ exp-de2ca         │ 08:51 PM     │    2 │ 2.2222 │  0.18385 │   2.1753 │ … │ 15     │ … │
│ │ ╟ caeaa60           │ 08:51 PM     │    1 │  2.291 │  0.14363 │   2.2434 │ … │ 15     │ … │
│ ├─╨ da4d04d           │ 08:51 PM     │    0 │ 2.3642 │  0.11472 │   2.3153 │ … │ 15     │ … │
└───────────────────────┴──────────────┴──────┴────────┴──────────┴──────────┴───┴────────┴───┘
```

> For `exp apply` it should probably accept any commit, but we have to determine the behavior for `exp run` in that case.

Right now the behavior seems correct except ambiguous-commit issue.

In general, if all the exp-name args can accept corresponded sha-s then we might hide the generated exp names for regular experiments.

**Comment by pmrowla** [comment]
> Could you please double-check this? To me, it seems to work from sha, not head. It looks like correct behavior.
> 
> ```shell
> $ vi params.yaml    # MODIFY PARAMS
```

This parameter change is what causes the new branch from sha

**Comment by dmpetrov** [comment]
> This parameter change is what causes the new branch from sha

Without modifications, "jumping" to the head/latest checkpoint makes total sense and also seems like correct behavior. So far, I don't see any issues with using/prioritizing sha-s. Please correct me if I missed something.

Messages can be improved a bit (minor issue):
1. No edits: `Existing checkpoint experiment 'exp-a9477' will be resumed`. It can be `Resuming from the latest checkpoint of 'exp-a9477'` (action word first, mention latest/head, skip word "experiment" - `exp-` prefix is explainable)
2. After edits: `Modified checkpoint experiment based on 'exp-a9477' will be created`. Can be: `Creating new 'exp-57b0b' based on the modified checkpoint from 'exp-a9477'` (mention new exp id)

Also, I'm having another issue. I'd appreciate it if you can explain: 
```
$ dvc exp apply 6a5ace6
$ dvc exp run
Existing checkpoint experiment 'exp-a9477' will be resumed
...
Updating lock file 'dvc.lock'
ERROR: Error running 'checkpoint' task, 'stage: 'train'' will be aborted - Failed to set 'refs/exps/10/dd6159b1946c31fb4c496415287f7cad5fcacc/exp-a9477'
```

**Comment by pmrowla** [comment]
> Without modifications, "jumping" to the head/latest checkpoint makes total sense and also seems like correct behavior. So far, I don't see any issues with using/prioritizing sha-s. Please correct me if I missed something.

The issue is regarding how to handle the ambiguous case, when there is more than one possible head/checkpoint to jump to from the given commit.

> Also, I'm having another issue. I'd appreciate it if you can explain

This looks like a bug, but it is hard to tell without the verbose output

**Comment by dmpetrov** [comment]
> > So far, I don't see any issues with using/prioritizing sha-s. Please correct me if I missed something.
> 
> The issue is regarding how to handle the ambiguous case

Is there a way to solve it? Based on my understanding (please correct me if I'm wrong), it comes down to the question - which of the exps DVC should use as a parent (and then decide if a "jump" to the head is needed or new exp is required). In any case, the exp when the initial checkpoint was created can be used. It supposes to be the exp with no modifications from the checkpoint up to its head.

From the example above it is `exp-8b809` for target checkpoint `c9cd463` but not `exp-70928` when modification `e9ebc70` can be recognized between `c9cd463` and exp head.

> This looks like a bug, but it is hard to tell without the verbose output

@pmrowla I'll try to reproduce it and send it directly to you.

**Comment by pmrowla** [comment]
> Is there a way to solve it? Based on my understanding (please correct me if I'm wrong), it comes down to the question - which of the exps DVC should use as a parent (and then decide if a "jump" to the head is needed or new exp is required). In any case, the exp when the initial checkpoint was created can be used. It supposes to be the exp with no modifications from the checkpoint up to its head.

There is not really any straightforward way for us to determine which experiment is the "original" and which is the "branched with modifications".

**Comment by dmpetrov** [comment]
We had a call with @pmrowla yesterday. Brief summary: today we cannot unambiguously identify which branch an experiment belongs to. This is where the issue `Ambiguous commit 'c9cd463' belongs to multiple experiment branches` comes from. It is not possible to identify because DVC cannot distinguish if the next commit is a result of user modification of the previous one or it is just a previous checkpoint. Some heuristic might help to identify but it will be computationally expensive. Today we use checkpoint (commit) creation time to identify the original experiment (with no modification). But there is a risk that this approach is not stable in a parallel environment.

To my mind, the parallel environment risk applies only to modified commits while the original one still should be created before any modification commit. So, we can safely use this timestamp-based logic to define the original checkpoint/commit for `dvc exp apply` to avoid the Ambiguous-commit issue. @pmrowla WDYH?

**Comment by pmrowla** [comment]
So from the ambiguous commit in the case where we have possible branches to choose from, we would walk to the commit where the branch fork happens, and then select the branch with the "oldest" (by timestamp) child from that fork, correct?

I think that should probably work for us, or at least right now I can't think of an edge case where it would not work.

I think it may still end up being a relatively expensive computation in situations where we have a lot of checkpoint branches containing a lot of modifications (basically where we have a large & complex commit tree w/a large # of possible choices), but that will require some testing to verify.

@dmpetrov

**Comment by dmpetrov** [comment]
> "oldest" (by timestamp) child from that fork, correct?

Yes. The modified version can happen only after the original one.

> I think it may still end up being a relatively expensive computation in situations 

I thought you are already doing time-based sorting for the table 😄 

> but that will require some testing to verify.

yep. Let's use it for now.

**Comment by pmrowla** [comment]
> I thought you are already doing time-based sorting for the table 😄

The current table just displays them in whatever order git happens to return them to us, which usually appears to be by timestamp but I don't think it's actually guaranteed?

**Comment by dberenbaum** [comment]
What was wrong with the original suggestion from @pmrowla?

> However, what we should maybe be doing here is assuming that the user intends to apply the intermediate commit and then start a new branch from that point? (this discussion should probably be moved into a new issue)

After `dvc exp apply c9cd463`, the next step would be to do some modification and then `dvc exp run`, which would generate a new modified experiment. The only time I can think of when the original experiment would matter is if no modification is made before `dvc exp run`, which is probably not an intended workflow since it would duplicate existing checkpoints.

**Comment by dmpetrov** [comment]
> After `dvc exp apply c9cd463`

@dberenbaum I'm not sure I understand this idea 😄  I'd appreciate it if you can clarify. But that the problem - `Apply` does not work in some cases due to the ambiguous-commit issue.

**Comment by dberenbaum** [comment]
I don't want to confuse by trying to explain what I think @pmrowla meant 🤣 , so maybe he can chime in first and I'll follow up if needed.

**Comment by dmpetrov** [comment]
> The current table just displays them in whatever order git happens to return them to us, which usually appears to be by timestamp but I don't think it's actually guaranteed?

@pmrowla well... It looks like an appealing idea to move the ambiguous-commit issue to Git's shoulders 😃 to avoid the computation and implementation overhead.

Is there a way to find out if Git does the time sorting? Even if we cannot find it, we can take the risk and use Git's result until we find when it does not work. @pmrowla what do you think?

**Comment by pmrowla** [comment]
> Is there a way to find out if Git does the time sorting? Even if we cannot find it, we can take the risk and use Git's result until we find when it does not work. @pmrowla what do you think?

It's just an implementation issue, when we are using multiple git backends (CLI git, dulwich, libgit2) they each have their own implementation and could return things in different orders.

> @dberenbaum I'm not sure I understand this idea 😄 I'd appreciate it if you can clarify. But that the problem - Apply does not work in some cases due to the ambiguous-commit issue.

The original suggestion was to never resume runs when we apply an ambiguous commit. We would just treat it as if the user always made a modification after this kind of apply, and start a new checkpoint "branch" from that point on the next `exp run` (even if the user did not actually modify any files). So rather than trying to determine which branch tip we have to jump to (and then continue), we just always create a new checkpoint branch entirely.

@dmpetrov

**Comment by dmpetrov** [comment]
> We would just treat it as if the user always intends to make a modification after this kind of apply, and start a new checkpoint "branch" from that point on the next `exp run`

@pmrowla thank you for the clarification. It would be great to keep a clear line of checkpoints if possible even after interruptions. Of course, we can relax the requirement to the clear line if the overhead is huge.

**Comment by dberenbaum** [comment]
> ```shell
> ```shell
> $ dvc exp show --no-pager
> ┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━┓
> ┃ Experiment            ┃ Created      ┃ step ┃     loss ┃ accuracy ┃ val_loss ┃ val_accuracy ┃ epochs ┃ lr      ┃ … ┃
> ┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━┩
> │ workspace             │ -            │    7 │ 0.054196 │  0.98373 │ 0.067046 │       0.9784 │ 15     │ 0.0005  │ … │
> │ master                │ Mar 30, 2021 │    - │        - │        - │        - │            - │ 15     │ 2.1e-07 │ … │
> │ │ ╓ exp-70928         │ 07:47 PM     │    7 │ 0.054196 │  0.98373 │ 0.067046 │       0.9784 │ 15     │ 0.0005  │ … │
> │ │ ╟ 47b34d6           │ 07:47 PM     │    6 │ 0.074431 │  0.97763 │ 0.076507 │        0.975 │ 15     │ 0.0005  │ … │
> │ │ ╟ 5471a57           │ 07:47 PM     │    5 │  0.11097 │  0.96748 │ 0.088773 │       0.9729 │ 15     │ 0.0005  │ … │
> │ │ ╟ e9ebc70 (7af692d) │ 07:47 PM     │    4 │  0.25394 │  0.92745 │  0.12504 │       0.9632 │ 15     │ 0.0005  │ … │
> │ │ ╓ exp-8b809         │ 07:47 PM     │    3 │   2.1693 │   0.2806 │   2.1272 │         0.35 │ 15     │ 2.1e-07 │ … │
> │ │ ╟ c9cd463           │ 07:47 PM     │    2 │   2.2433 │  0.21087 │   2.2005 │       0.2653 │ 15     │ 2.1e-07 │ … │
> │ │ ╟ a7a30b4           │ 07:47 PM     │    1 │   2.3233 │  0.15603 │   2.2794 │       0.1925 │ 15     │ 2.1e-07 │ … │
> │ ├─╨ a1fddb3           │ 07:47 PM     │    0 │   2.4123 │   0.1174 │   2.3657 │       0.1307 │ 15     │ 2.1e-07 │ … │
> └───────────────────────┴──────────────┴──────┴──────────┴──────────┴──────────┴──────────────┴────────┴─────────┴───┘
> ```
> 
> 
> ```shell
> $ dvc exp apply c9cd463
> ERROR: Ambiguous commit 'c9cd463' belongs to multiple experiment branches.
> ```

@dmpetrov After `dvc exp apply c9cd463`, wouldn't you always be making some modification, which will create a new experiment anyway? Otherwise, you would be duplicating step 3 of `exp-8b809`.

It seems this is only an issue if you check out the tip of the experiment branch, in which case you could do`dvc exp apply exp-8b809` instead.

By the way, getting back to the table format, `e9ebc70 (7af692d)` is kind of confusing in the table above since `7af692d` doesn't show up in any other line. It looks like it refers back to the tip of `exp-8b809`, but that gets replaced with `exp-8b809`.

**Comment by dmpetrov** [comment]
> @dmpetrov After `dvc exp apply c9cd463`, wouldn't you always be making some modification, which will create a new experiment anyway?

@dberenbaum good point! In some cases, you do `apply xxxx & run` but it didn't work and you would like to continue the training in "old" checkpoint. However, the next training should start from head not from `xxxx`. 

@pmrowla if we drop the support for straight line in no-modified training. Can we still have that if we train from head? Like:

```
$ dvc exp apply 123456
# no modifiactions
$ dvc exp run # Creates a new experiment in the tree.
# I'd like to continue to train from head:
$ dvc exp continue exp-7654 # only exp- is supported, not sha (or sha from head)
```

> By the way, getting back to the table format, `e9ebc70 (7af692d)` is kind of confusing in the table above

Yep. It is another reason why sha-s are more convinient.

**Comment by dberenbaum** [comment]
> It seems that creating a new experiment after apply is only an issue if you check out the tip of the experiment branch, in which case you could do`dvc exp apply exp-8b809` instead.

> @dberenbaum good point! In some cases, you do `apply xxxx & run` but it didn't work and you would like to continue the training in "old" checkpoint. However, the next training should start from head not from `xxxx`.

I think we are saying the same thing here, so seems like we just need @pmrowla to confirm whether it will work as expected for this scenario. I was thinking slightly differently only in terms of syntax to avoid needing to introduce `continue` again:

```console
$ dvc exp apply 123456
# no modifications
## If this is not the head/tip of an exp, exp run would ideally throw an error since the checkpoint should already exist.
$ dvc exp run # Creates a new experiment in the tree.
# I'd like to continue to train from head:
$ dvc exp apply exp-7654 # only exp- is supported, not sha
# no modifications
$ dvc exp run # Adds checkpoints to exp-7654
```

**Comment by pmrowla** [comment]
> ```
> $ dvc exp apply 123456
> # no modifications
> ## If this is not the head/tip of an exp, exp run would ideally throw an error since the checkpoint should already exist.
> $ dvc exp run # Creates a new experiment in the tree.
> # I'd like to continue to train from head:
> $ dvc exp apply exp-7654 # only exp- is supported, not sha
> # no modifications
> $ dvc exp run # Adds checkpoints to exp-7654
> ```

From the engineering perspective, doing it this way would be preferable over traversing git history (in order to find the "oldest" branch).

We also don't need to have the limitation here:
> ```
> $ dvc exp apply exp-7654 # only exp- is supported, not sha
> # no modifications
> $ dvc exp run # Adds checkpoints to exp-7654
> ```
We can allow the sha for `exp-7654` here - determining that a sha is an exact match for an `exp-` head is a cheap operation that does not actually require traversing any git history.
